server/src/main/java/org/jetbrains/bsp/bazel/server/bsp/BazelBspServerBuildManager.java

```diff
@@ -64,13 +64,32 @@ public BazelBspServerBuildManager(
     return bazelBspQueryManager.getSourceItems(rule, label);
   }
 
-  public String getSourcesRoot(String uri) {
-    List<String> root =
-        Constants.KNOWN_SOURCE_ROOTS.stream().filter(uri::contains).collect(Collectors.toList());
-    return bazelData.getWorkspaceRoot()
-        + (root.size() == 0
-            ? ""
-            : uri.substring(1, uri.indexOf(root.get(0)) + root.get(0).length()));
+  public String getSourcesRoot(URI sourceUri) {
+    Path sourcePath = Paths.get(sourceUri);
+    FileSystem fs = FileSystems.getDefault();
+    PathMatcher sourceRootPattern = fs.getPathMatcher(
+            "glob:**/"
+                    + "{main,test,tests,src,3rdparty,3rd_party,thirdparty,third_party}/"
+                    + "{resources,scala,java,kotlin,jvm,proto,python,protobuf,py}");
+    PathMatcher defaultTestRootPattern = fs.getPathMatcher("glob:**/{test,tests}");
+    Path sourceRootGuess = null;
+    for (PathMatcher pattern : new PathMatcher[]{sourceRootPattern, defaultTestRootPattern}) {
+      sourceRootGuess = approximateSourceRoot(sourcePath, pattern);
+      if (sourceRootGuess != null) break;
+    }
+    if (sourceRootGuess == null) {
+      return sourcePath.getParent().toString();
+    }
+    return sourceRootGuess.toAbsolutePath().toString();
```

Comment:
```suggestion
Optional<Path> patternPath = ImmutableList.of(sourceRootPattern, defaultTestRootPattern).stream()
      .map(pattern -> approximateSourceRoot(sourcePath, pattern))
      .filter(Optional::isPresent)
      .map(Optional::get)
      .findFirst()
      .map(Path::toAbsolutePath);

  return patternPath.orElse(sourcePath.getParent())
      .toString();
```

Reply:
I think that my variant is more readable, even though it uses null.

Reply:
i strongly disagree - at least for me it looks very messy

Reply:
Both are readable to me. I don't think we should make such style matters a big issue in accepting PRs ;)

Reply:
Refactored